Builtin thread-local object storage for AsyncTask

SOF3 opened the pull request:
@dktapps deleted my original pull request. Blame him and ask him for description.

SOF3 committed:
Added AsyncTask::fetchLocal()

Update AsyncTask.php

Fix syntax error

archie426 commented:
Go for it, nice to see that you are here now! Btw, did u leave the forums?

dktapps commented:
#blamedktapps
Sorry this has been such a pain in the butt.

SOF3 commented:
Rebased commits

Aericio commented:
@archie426 He got banned on forums lol

SOF3 commented:
@Aericio Please do not post comments irrelevant to this pull request here.

robske110 commented on src/pocketmine/scheduler/AsyncTask.php:
```diff
@@ -27,7 +27,7 @@
  * Class used to run async tasks in other threads.
  *
- * WARNING: Do not call PocketMine-MP API methods, or save objects from/on other Threads!!
+ * WARNING: Do not call PocketMine-MP API methods, or save objects (and arrays cotaining objects) from/on other Threads!!
```
love this :P

remotevase replied:
Lol hilarious

SOF3 replied:
@remotevase please stop spamming on our repo.

remotevase replied:
Okay, but expect to see more pull requests from me that actually contribute CODE not just templates for issues and pull requests.

SOF3 replied:
@remotevase Please immediately **stop** using pull requests and issues as chatroom or forums, or respective action will be taken.

robske110 reviewed (approved):
Awesome :)

Aericio commented:
> @Aericio Please do not post comments irrelevant to this pull request here. If you continue to post improperly, you may be blocked from @pmmp.

Mind deleting that now? How relevant.

robske110 commented:
I could try to test this, if you want
-?

robske110 commented:
The commit name seems to be not very descriptive

SOF3 commented:
@robske110 that's all that this pull request is about. Other changes are just to assist this addition. You don't want to store a value; you just want to retrieve a value, and you do so by storing it first, along with other things like a storage to put these values at. But the main point is that you can use `AsyncTask;:fetchLocal`.

robske110 commented:
well, still looks weird for me, but i guess ok

SOF3 committed:
Merge branch 'master' into al

SOF3 commented:
Tested with this script plugin:

```php
<?php

/**
 * @name AsyncTest
 * @author SOFe
 * @main AsyncTest\AsyncTest
 * @api 2.1.0
 * @version 1.0.0
 */

namespace AsyncTest;

class AsyncTest extends \pocketmine\plugin\PluginBase {
        public function onEnable(){
                $asyncTask = new class($this) extends \pocketmine\scheduler\AsyncTask{
                        public function onRun(){
                                sleep(5);
                        }

                        public function onCompletion(\pocketmine\Server $server){
                                $server->getLogger()->info('$asyncTask');
                                var_dump(get_class($this->fetchLocal($server)));
                        }
                };
                $this->getServer()->getScheduler()->scheduleAsyncTask($asyncTask);

                $asyncTask2 = new class($this) extends \pocketmine\scheduler\AsyncTask{
                        public function onRun(){
                                sleep(10);
                        }

                        public function onCompletion(\pocketmine\Server $server){
                                $server->getLogger()->info('$asyncTask2');
                                // var_dump(get_class($this->fetchLocal($server)));
                        }
                };
                $this->getServer()->getScheduler()->scheduleAsyncTask($asyncTask2);
        }
}
```

Note that anonymous classes are used in this script, and the values to store for an AsyncTask is the `$this` in the `new class($this)` part.

Output with `./start.sh --debug.level=2`:

```php
[12:07:07] [Server thread/INFO]: Enabling AsyncTest v1.0.0
# Irrelevant lines skipped
[12:07:07] [Server thread/INFO]: Done (1.414s)! For help, type "help" or "?"
[12:07:12] [Server thread/INFO]: $asyncTask
~\plugins\async.php:22:
string(19) "AsyncTest\AsyncTest"
[12:07:22] [Server thread/INFO]: $asyncTask2
[12:07:22] [Server thread/DEBUG]: AsyncTask class@anonymous~\plugins\async.php0000006F77B8036A stored local complex data but did not remove them after completion
```

SOF3 commented on src/pocketmine/scheduler/ServerScheduler.php:
```diff
@@ -91,6 +95,59 @@ public function scheduleAsyncTaskToWorker(AsyncTask $task, $worker){
+	 */	
+	public function removeLocalComplex(AsyncTask $for) : bool{
+		if(isset($this->objectStore[$for])){
+			Server::getInstance()->getLogger()->debug("AsyncTask " . get_class($for) . " stored local complex data but did not remove them after completion");
```
Question: Should this be at the WARN level instead?

SOF3 committed:
Updated PHP doc

SOF3 commented:
The pull request is ready for merge. Apart from the question about WARN vs DEBUG above, it is OK for merging. Please review this pull request if you find no problems in the API design.

robske110 commented:
I think NOTICE would be the right level.

SOF3 committed:
Changed log level DEBUG to NOTICE

Changed log level of potential memory leak from DEBUG to NOTICE.

This is really wrong. It should be shown even on production servers.

robske110 reviewed (approved):
Looks good now, I'd say ready for merge 👍

dktapps commented on src/pocketmine/scheduler/AsyncTask.php:
```diff
@@ -42,6 +42,29 @@
+	 * Constructs a new instance of AsyncTask. Subclasses don't need to call this constructor unless an argument is to be passed. ONLY construct this class from the main thread.
+	 * <br>
+	 * If an argument is passed into this constructor, it will be stored in a thread-local storage (in ServerScheduler), which MUST be retrieved through {@link #fetchLocal} when {@link #onCompletion} is called.
+	 * Otherwise, a DEBUG level message will be raised and the reference will be removed after onCompletion exits.
```
Didn't you change this?

SOF3 committed:
Update AsyncTask.php

dktapps reviewed (approved):
Discussed in detail.
